ouroboros-consensus-protocol/src/ouroboros-consensus-protocol/Ouroboros/Consensus/Protocol/Ledger/HotKey.hs

```diff
@@ -126,19 +135,58 @@ data HotKey c m = HotKey {
       --
       -- When the key cannot evolve anymore, we poison it.
       evolve     :: Absolute.KESPeriod -> m KESEvolutionInfo
+
       -- | Return 'KESInfo' of the signing key.
     , getInfo    :: m KESInfo
-      -- | Return 'True' when the signing key is poisoned because it expired.
+
+      -- | Return the 'OCert' corresponding to the KES signing key, if any.
+    , getOCertMaybe :: m (Maybe (OCert.OCert c))
+
+      -- | Check whether a valid KES signing key exists. "Poisoned" means no
+      -- key exists; reasons for this could be:
+      -- - no signing key has been set yet
+      -- - the signing key has been explicitly erased ('forget')
+      -- - the signing key has been evolved past the end of the available
+      --   evolutions
     , isPoisoned :: m Bool
+
       -- | Sign the given @toSign@ with the current signing key.
       --
       -- PRECONDITION: the key is not poisoned.
       --
       -- POSTCONDITION: the signature is in normal form.
     , sign_      :: forall toSign. (SL.KESignable c toSign, HasCallStack)
-                 => toSign -> m (SL.SignedKES c toSign)
+                 => toSign
+                 -> m (SL.SignedKES c toSign)
+
+      -- | Securely erase the key and release its memory.
+    , forget :: m ()
+
+      -- | Set a new sign key.
+    , set :: OCert.OCert c
+             -- ^ The new OCert
+          -> SL.SignKeyKES c
+             -- ^ The new KES key
+          -> Word
+             -- ^ The new KES key's current evolution
+          -> Absolute.KESPeriod
+             -- ^ Start period (relative to the KES key's 0th evolution)
+          -> m ()
+
+      -- | Release any resources held by the 'HotKey'. Must be run exactly once
+      -- per 'HotKey'.
+    , finalize :: m ()
```

Comment:
Hoy is `finalize` different from `forget`? Or how do those relate, should one call both?

Reply:
`forget` erases the key itself, `finalize` deals with any other resources that may be associated with the HotKey.

You just pointed me at a bug, though: we provide a `forget` method that erases the key, but we are not actually using it anywhere, so we actually leak mlocked memory on shutdown.

Reply:
Fixed that by renaming `finalize` to `finalize_`, and exporting a wrapper, `finalize`, that calls both.